Source/santad/EventProviders/SNTEndpointSecurityTamperResistance.mm

```diff
@@ -85,6 +86,51 @@ - (void)handleMessage:(Message &&)esMsg
       break;
     }
 
+    case ES_EVENT_TYPE_AUTH_SIGNAL: {
+      // Only block signals sent to us and not from launchd.
+      if (audit_token_to_pid(esMsg->event.signal.target->audit_token) != getpid() ||
+          audit_token_to_pid(esMsg->process->audit_token) == 1) {
+        result = ES_AUTH_RESULT_ALLOW;
+        break;
+      }
+      LOGW(@"Preventing attempt to kill Santa daemon");
+      result = ES_AUTH_RESULT_DENY;
+      break;
+    }
+
+    case ES_EVENT_TYPE_AUTH_EXEC: {
+      // When not running a debug build, prevent attempts to kill Santa
+      // by launchctl commands.
```

Comment:
We chatted briefly about this yesterday, but here's a quick sorta-pseudo-code approach I was mentioning. I don't expect this to be a hot path at all, but inverting these checks will reduce the number of string comparisons done:

```
es_auth_result_t ValidateLaunchctl(const Message &esMsg) {
  es_string_token_t exec_path = esMsg->event.exec.target->executable->path;
  if (strncmp(exec_path.data, "/bin/launchctl", exec_path.length) != 0) {
    return ES_AUTH_RESULT_ALLOW;
  }

  // Ensure at least 1 argument is provided to the command
  std::shared_ptr<EndpointSecurityAPI> esApi = esMsg.ESAPI();
  uint32_t argCount = esApi->ExecArgCount(&esMsg->event.exec);
  if (argCount < 2) {
    return ES_AUTH_RESULT_ALLOW;
  }

  // Check for some allowed subcommands
  es_string_token_t arg = esApi->ExecArg(&esMsg->event.exec, 1);
  if (arg.data in ["list" || "print"]) {
    return ES_AUTH_RESULT_ALLOW;
  }

  // Check whether com.google.santa.daemon is in the argument list.
  // launchctl no longer accepts PIDs to operate on.
  for (int i = 2; i < argCount; i++) {
    es_string_token_t arg = esApi->ExecArg(&esMsg->event.exec, i);
    if (strnstr(arg.data, "com.google.santa.daemon", arg.length) != NULL) {
      return ES_AUTH_RESULT_DENY;
    }
  }

  return ES_AUTH_RESULT_ALLOW;
}
```

Reply:
To make sure I'm reading this right at a high level the flow is:

```mermaid
graph TD
A[Execution ESMsg is received] --> B{Is the command /bin/launchctl?}
B --> |No| X[Return ES_AUTH_RESULT_ALLOW]
B --> |Yes| C{Does this have at least 1 argument?}
C --> |No|X
C --> |Yes|D{Is the first argument\n a safe subcommand?}
D --> |Yes|X
D --> |No|E{Is the command targeting\n the Santa daemon \naka com.google.santa.daemon?}
E --> |Yes|Y[Return ES_AUTH_RESULT_DENY]
E --> |No|X
```

I imagine that rather than checking the path in the binary we'd use the Signing ID so that someone can't copy the binary or other trickery.

Reply:
Flow LGTM.

> I imagine that rather than checking the path in the binary we'd use the Signing ID so that someone can't copy the binary or other trickery.

Fine with checking "platform binary + SID", but also it shouldn't much matter either way. You can't hard link this binary, symlinks would resolve to the real path in ES, and launch constraints ___should___ prevent any other copy trickery.

Reply:
>  launch constraints should prevent any other copy trickery.

Yeah, tbh I'm not entirely sure what the limits are with launch constraints but given that we have a strong primitive like code signing seems like an easy way to be sure.

Reply:
Using the signing ID would mean unmuting all paths, which makes this _much_ more expensive. Given the limitations on copying/linking launchctl, I think this is good enough as-is.